type-check.ts

```diff
@@ -63,19 +73,81 @@ export type TypeError = {
   message: string
 }
 
-export function equalType(t1: Type, t2: Type) {
+// combine the elements of two arrays into an array of tuples.
+// DANGER: behaves weirdly with arrays of different length.
+function zip<A, B>(l1: Array<A>, l2: Array<B>) : Array<[A, B]> {
+  return l1.map((el, i) => [el, l2[i]]); 
+}
+
+export function equalType(env: GlobalTypeEnv, t1: Type, t2: Type) {
+  // If either of t1 or t2 are type-variables, then the actual type
+  // that they are currently instantiated to needs to be fetched from
+  // the environment before comparing the types.
+  if(t1.tag === "typevar") {
+    t1 = env.typevars.get(t1.name)[2];
+  }
+
+  if(t2.tag === "typevar") {
+    t2 = env.typevars.get(t2.name)[2];
+  }
+  
   return (
     t1 === t2 ||
-    (t1.tag === "class" && t2.tag === "class" && t1.name === t2.name)
+    (t1.tag === "class" && t2.tag === "class" && t1.name === t2.name && equalTypeParams(env, t1.params, t2.params))
   );
 }
 
-export function isNoneOrClass(t: Type) {
-  return t.tag === "none" || t.tag === "class";
+// Check if a list of type-parameters are equal.
+export function equalTypeParams(env: GlobalTypeEnv, params1: Type[], params2: Type[]) : boolean {
+  if(params1.length !== params2.length) {
+    return false;
+  }
+
+  return zip(params1, params2).reduce((isEqual, [p1, p2]) => {
+    return isEqual && equalType(env, p1, p2);
+  }, true);
+}
+
+export function isNoneOrClass(env: GlobalTypeEnv, t: Type) {
+  if(t.tag === "typevar") {
+    // If t is a type-variables, then the actual type
+    // that it is currently instantiated to needs to be fetched from
+    // the environment.
+    t = env.typevars.get(t.name)[2];
+  }
+
+  // TODO: decide if something with type-parameters is considered a class
+  return t.tag === "none" || (t.tag === "class" && t.name !== UNCONSTRAINED);
+}
+
+export function getClassName(env: GlobalTypeEnv, t: Type): string {
+  if(t.tag === "typevar") {
+    // If t is a type-variables, then the actual type
+    // that it is currently instantiated to needs to be fetched from
+    // the environment.
+    t = env.typevars.get(t.name)[2];
+  }
+
+  if(t.tag !== "class" || t.name === UNCONSTRAINED) {
+    return undefined;
+  }
+
+  return t.name;
 }
 
 export function isSubtype(env: GlobalTypeEnv, t1: Type, t2: Type): boolean {
-  return equalType(t1, t2) || t1.tag === "none" && t2.tag === "class" 
+  // If either of t1 or t2 are type-variables, then the actual type
+  // that they are currently instantiated to needs to be fetched from
+  // the environment before comparing the types.
```

Comment:
I'm confused about how we know which types they are instantiated to. That phrase doesn't make sense when, say, the whole program is just a generic class.

Reply:
In the current implementation where we support Generic Classes only, our typechecking process for each class is -
 1. Replace all entities "wrongly" labelled as "class" but are infact type-variables with the "typevar" tag. These are all the sources of genericness within each class.
 2. Cycle through all possible combinations of assignments for the type-parameters for the class, save the assignment in the environment and run the existing `tcClass`function.

To answer your questions - 
1. We know the types they are assigned to by looking at the environment. Based on the above implementation, we should never encounter something with the type tag "typevar" outside of the context of a Generic Class being typechecked using tcClass with a particular assignment of type-variables in the environment.
2. If the whole program is a generic class, it is not a problem. The assignment in the environment is based on the constraints for the Type-Variables and are independent of the actual instantiations for the type-parameter made during the program.